File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-3.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} p
     * @param {TreeNode} q
     * @return {TreeNode}
     */
    lowestCommonAncestor(
        root: TreeNode | null,
        p: TreeNode | null,
        q: TreeNode | null,
    ) {
        let cur = root;
        while (cur) {
            if(cur.val < p.val && cur.val < q.val) { 
                cur = cur.right;
            } else if (cur.val > p.val && cur.val > q.val) { 
                cur = cur.left;
            } else {
                return cur; 
            }
        }
        return null;
    }
}

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        pathSum = float("-inf")
        def dfs(node):
            nonlocal pathSum
            if not node:
                return 0
            leftMax = max(0, dfs(node.left))
            rightMax = max(0, dfs(node.right))
            # traversing the node
            currentPath = node.val + leftMax + rightMax

            pathSum = max(pathSum, currentPath)

            return node.val + max(leftMax, rightMax)
        dfs(root)

        return pathSum

        